### DISCUSSION TITLE
How to write custom sql for database?

### DISCUSSION THREAD

**Discussion by plantshopping** [pocketbase/pocketbase#5987] [Q&A]
I am trying to write some custom sql in my next js app for a chat application to get preview of all messages a user is involved in, but how do I do it with the pocketbase database?

```
   $app.dao().db()
      .select(
        `
        m1.id AS message_id,
        m1.content AS latest_message,
        m1.sender AS sender_id,
        m1.recipient AS recipient_id,
        m1.created_at AS sent_at,
        CASE
            WHEN m1.sender = {:current_user} THEN m1.recipient
            ELSE m1.sender
        END AS chat_partner_id
    `
      )
      .from("messages AS m1")
      .where(
        $dbx.exp(`
            m1.created_at = (
                SELECT MAX(m2.created_at)
                FROM messages m2
                WHERE
                    (m1.sender = m2.sender AND m1.recipient = m2.recipient) OR
                    (m1.sender = m2.recipient AND m1.recipient = m2.sender)
            )
        `)
      )
      .andWhere(
        $dbx.exp("{:current_user} IN (m1.sender, m1.recipient)", {
          current_user: user.id,
        })
      )
      .orderBy("m1.created_at DESC")
      .all(result);
```

But what is $app?

**Comment by shajidhasan** [comment]
Why can't you use the PocketBase [js-sdk](https://github.com/pocketbase/js-sdk) for this? Maybe something like `pb.collection('messages').getList(1, 100, {filter: `sender="${sender_id}"`, sort: '-created'})`.

And you can get access to `$app` in the PocketBase hooks. [Learn more here](https://pocketbase.io/docs/js-overview/).

**Reply by plantshopping** [reply]
I'm not sure how to write it using the PocketBase js-sdk

**Reply by chnfyi** [reply]
I think it means that you can create a [View collection](https://pocketbase.io/docs/collections/#view-collection) in the UI dashboard or use the js SDK to retrieve data with: `pb.collection('messages').getFullList("sender=${sender_id}")`

**Reply by plantshopping** [reply]
I think my query is too complex to just write with the pocketbase js-sdk so i would like to write it with raw sql

**Reply by chnfyi** [reply]
Based on your goal _"to get preview of all messages a user is involved in"_ and your example, i think you can achieve it in a simpler way with the js SDK as previously suggested:

```javascript
const id = pb.authStore.record?.id
const messages = await pb.collection('messages').getFullList({
    filter: `sender=${id} || recipient=${id}`,
    sort: '-created'
})
```

**Reply by plantshopping** [reply]
It's not as simple as that, a user can have hundreds of messages with different users, I need to get the latest message for each conversation to show in a preview panel (much like how facebook or instagram does previews)

**Reply by KelvinConnor** [reply]
If you need to use custom SQL, you just need to create a custom Route. If using JS, basically you just create a javascript file with a certain extension and deploy it on the server with your pocketbase instance. Then you can call the Route from your application.

To create a Route:
https://pocketbase.io/docs/js-overview/#javascript-engine
https://pocketbase.io/docs/js-routing/#registering-new-routes

Calling a Route:
https://pocketbase.io/docs/js-routing/#sending-request-to-custom-routes-using-the-sdks

The initial $app code you included would be run in that route. Be sure to read all the doco under "Extend with Javascript" section. There is information about how to return the data, security, and logging that you will want to understand.

**Reply by plantshopping** [reply]
Is there a complete sample for how to do write a custom sql query with a new route?

**Reply by KelvinConnor** [reply]
I am not sure there is a full example in the documentation (in one place). There are some query examples under Executing Queries  - https://pocketbase.io/docs/js-database/#executing-queries

If you create a file called something like 'custom_endpoints.pb.js' in the /pb_hooks directory, you should be able to create a custom SQL route in it like this:

PLEASE NOTE THIS IS USING THE v0.23 VERSION OF PB AND IS ONLY A BAREBONES EXAMPLE

```
routerAdd("GET", "/myCustomRoute", (c) => {
    //use arrayOf if the response is going to be a list of records
    const result = arrayOf(new DynamicModel({
        // describe the shape of the data (used also as initial values)
        "username": "",
        "name": "", 
        "email": "",
    }));
    const userId = c.requestInfo().auth.id ?? 'MISSINGAUTH!';
    try {
        $app.db()
            .newQuery("SELECT username, name, email FROM users WHERE id = '" + userId + "'")
            .all(result)
        console.log("/myCustomRoute: Found " + result.length + " results")
    } catch (error) {
        console.error("/myCustomRoute: ERROR " + error);
        $app.logger().error("/myCustomRoute: JS ERROR " + error);
    }
    return c.json(200, result)
}, $apis.requireAuth()) 
```

Then you can call it from the client SDK.  Hope that helps.

**Reply by chnfyi** [reply]
While it still may not fit your use case, you can add a conversation id which each message refers, then with js SDK:

```javascript
const id = pb.authStore.record?.id

// Retrieve the 30 latest messages sorted by conversation
const messages = await pb.collection('messages').getList(1, 30, {
    filter: `sender=${id} || recipient=${id}`,
    sort: '-created, conversationId'
})
``` 

Otherwise, you can write plain SQLs (or use a [query builder](https://pocketbase.io/docs/js-database/#query-builder) if it's more convenient for you) in a custom route as previously suggested.
